platform/mbed_error.h

```diff
@@ -23,13 +22,793 @@
+    ENETUNREACH                101      Network is unreachable 
+    ENETRESET                  102      Network dropped connection because of reset 
+    ECONNABORTED               103      Software caused connection abort 
+    ECONNRESET                 104      Connection reset by peer 
+    ENOBUFS                    105      No buffer space available 
+    EISCONN                    106      Transport endpoint is already connected 
+    ENOTCONN                   107      Transport endpoint is not connected 
+    ESHUTDOWN                  108      Cannot send after transport endpoint shutdown 
+    ETOOMANYREFS               109      Too many references: cannot splice 
+    ETIMEDOUT                  110      Connection timed out 
+    ECONNREFUSED               111      Connection refused 
+    EHOSTDOWN                  112      Host is down 
+    EHOSTUNREACH               113      No route to host 
+    EALREADY                   114      Operation already in progress 
+    EINPROGRESS                115      Operation now in progress 
+    ESTALE                     116      Stale NFS file handle 
+    EUCLEAN                    117      Structure needs cleaning 
+    ENOTNAM                    118      Not a XENIX named type file 
+    ENAVAIL                    119      No XENIX semaphores available 
+    EISNAM                     120      Is a named type file 
+    EREMOTEIO                  121      Remote I/O error 
+    EDQUOT                     122      Quota exceeded 
+    ENOMEDIUM                  123      No medium found 
+    EMEDIUMTYPE                124      Wrong medium type 
+    ECANCELED                  125      Operation Canceled 
+    ENOKEY                     126      Required key not available 
+    EKEYEXPIRED                127      Key has expired 
+    EKEYREVOKED                128      Key has been revoked 
+    EKEYREJECTED               129      Key was rejected by service 
+    EOWNERDEAD                 130      Owner died 
+    ENOTRECOVERABLE            131      State not recoverable 
+    \endverbatim
+ *
+ *  @note
+ *  MbedOS System Error codes are defined using the macro DEFINE_SYSTEM_ERROR\n
+ *  For example DEFINE_SYSTEM_ERROR( INVALID_ARGUMENT ,1 ) macro defines the following values:\n
+ *      ERROR_CODE_INVALID_ARGUMENT = MBED_SYSTEM_ERROR_BASE+1\n
+ *      ERROR_INVALID_ARGUMENT = MAKE_MBED_ERROR(ERROR_TYPE_SYSTEM, MODULE_UNKNOWN, ERROR_CODE_INVALID_ARGUMENT)\n
+ *  Its effectively equivalent to:\n
+ *      ERROR_CODE_INVALID_ARGUMENT = 1\n
+ *      ERROR_INVALID_ARGUMENT = 0x80FF0001\n (Note that MODULE field is set to MODULE_UNKNOWN) 
+ *  New System Error codes should be defined using DEFINE_SYSTEM_ERROR macro and must have an unique error code value\n
+ *  passed as the second argument in the DEFINE_SYSTEM_ERROR macro.\n\n
+ *  Below are the Mbed System error codes and the description:
+ *  \verbatim 
+    UNKNOWN                    256      Unknown error 
```

Comment:
This is already pretty well established as -1. Can we keep -1 as an unknown error?

Reply:
I was thinking about it but -1 maps to Posix Error Code. So, mapping -1 again to unknown error maybe confusing. And it also doesn't align with our overall error code definition/format.

Reply:
I'm thinking more we should renumber the std errno codes (see above). This would free up -1 as general error.

Actually even if it conflicts, its just conflicting with EPERM (not EACCESS), which is pretty close to a generic error.

Reply:
The idea here is to include posix error codes as part of mbed error handling model so the modules already using posix still have the same meaning. The problem with conflicting error codes is that it becomes difficult if some one gets an error and wants to search source code for where that came from.

Reply:
There are far more modules returning -1 for errors than POSIX errors. Back in mbed 2 this was the correct thing to do:
https://os.mbed.com/users/mbed_official/code/SDFileSystem/file/8db0d3b02cec/SDFileSystem.cpp

The only thing worse than conflicting error codes are error codes with misleading names.

This also does not become a problem if we renumber the std errno codes.